### DISCUSSION TITLE
Uploaded files truncated to 16 megabytes

### DISCUSSION THREAD

**Discussion by jannefoo** [io7m-com/exfilac#28] [Q&A]
original file on phone: 17 797 262 bytes
file downloaded from aws s3 bucket: 16 777 216 bytes

I'm running 1.1.1 from f-droid on pixel 8a. 

from exfilac's logs, its seems the file is uploaded as two 8388608 byte chunks, but three would've been needed.

a bug or have I misconfigured something?

**Comment by jannefoo** [comment]
btw, exfilac's log also say that hashes match between local and s3 files, but how that can be when file sizes mismatch? 🤔

sub-chunk sized files are uploaded intact.

**Comment by io7m** [comment]
Very strange.

What's the S3 server provider in this case?

As you may know, we go through the Amazon AWS SDK. For files as large as you mention, this will be the multi-part upload code path. Essentially, this method:

https://github.com/io7m-com/exfilac/blob/17b07f4190601f271ff3ad9b7bbb4829544dc9f0/com.io7m.exfilac.s3_uploader.amazon/src/main/java/com/io7m/exfilac/s3_uploader/amazon/EFS3AMZUpload.kt#L276

You can see that we set the expected hash of the file as part of the metadata for the upload. This should mean (assuming that the S3 provider adheres to the spec) that you can't observe a file that exposes the hash value that `exfilac` reads back until that multi-part upload has been fully completed and the content verified on the server side.

I'm also seeing occasional odd `403` errors on a locally hosted MinIO server these past couple of days related to multi-part uploads, so I'm suspecting that something isn't quite right somewhere with respect to the underlying S3 client.

**Reply by jannefoo** [reply]
My host is Amazon S3.

**Comment by io7m** [comment]
Tracking this here: https://github.com/io7m-com/exfilac/issues/29

**Comment by io7m** [comment]
Thanks for reporting this. This turned out to be a pretty serious issue with multi-part size calculation that was somehow left out of testing. The latest commit should fix this, but I'm running it through a period of testing locally before a new release goes out.

**Comment by io7m** [comment]
Going to close this discussion now as I believe this is fixed.
